Improve child blocks API's and UI.

jorgefilipecosta opened the pull request:
## Description
This PR extends the blocks API with two new selectors getChildBlocks and hasChildBlocks.
A class is added for inserter items with children and the UI is changed using the new class. Now we have a special design in blocks with children (this design is in progress).
A new area that appears when we are inserting blocks inside blocks with children was created.

jorgefilipecosta committed:
Improve child blocks API's and UI.

jorgefilipecosta commented on blocks/store/selectors.js:
```diff
@@ -61,3 +62,39 @@ export function getDefaultBlockName( state ) {
+ */
+export const hasChildBlocks = createSelector(
+	( state, blockName ) => {
+		return getChildBlocks( state, blockName ).length > 0;
```
Nice catch @gziolo, I updated the code!

mtias committed:
Adjust visual design of stacked block.

mtias commented:
Closes #6998.

jorgefilipecosta commented:
Hi @gziolo the test case was fixed.
Thank @mtias for the design updates.

mtias committed:
Polish design of block cards in the inserter.

mtias commented:
Pushed some further tweaks to the design:

![image](https://user-images.githubusercontent.com/548849/40715130-e0f138ac-6404-11e8-9470-9df7d4d330d6.png)

@jasmussen I think the biggest thing missing is addressing the blurry icons :)

Commit:
Don't blur the icons.

jasmussen commented:
Nice work. I unblurred the icons :D

jorgefilipecosta commented:
I rebased this PR and it seems to be working fine.

gziolo commented on editor/components/inserter/menu.js:
```diff
@@ -176,6 +184,18 @@ export class InserterMenu extends Component {
+						<div className="editor-inserter__child-blocks">
+							<div className="editor-inserter__parent-block-header">
+								<div className="editor-inserter__parent-block-icon">
+									<BlockIcon icon={ rootBlockIcon } />
```
BlockIcon should be rendered only when the icon is set

jorgefilipecosta replied:
Normally we never do this check because BlockIcon handles it but in this case, we can do it because we have wrapping div for the icon.

gziolo commented on editor/components/inserter/menu.js:
```diff
@@ -176,6 +184,18 @@ export class InserterMenu extends Component {
 				/>
 
 				<div className="editor-inserter__results">
+					{ !! childItems.length &&
```
Can we extract this code to a new component to avoid confusing conditional checks for an icon and title? As far as I understand when there are child blocks then the icon and title must be set, right?

jorgefilipecosta replied:
Yes if we have childblocks the icon and title are set unless the block did not use a title/item, but that's part of the block responsibility.

jorgefilipecosta replied:
The child blocks rendering logic was extracted to a separate component.

gziolo commented on blocks/store/selectors.js:
```diff
@@ -61,3 +62,37 @@ export function getDefaultBlockName( state ) {
+ *
+ * @return {Array} Array of child block names.
+ */
+export const getChildBlockNames = createSelector(
```
This selector seems to be quite important so let’s add unit tests to avoid regressions in the future.

jorgefilipecosta replied:
Unit tests added 👍

gziolo commented on editor/components/inserter/menu.js:
```diff
@@ -109,9 +112,13 @@ export class InserterMenu extends Component {
+		const { items, rootChildBlocks } = this.props;
 		const filteredItems = searchItems( items, filterValue );
 
+		const childItems = rootChildBlocks ?
```
rootChildBlocks is always an array, so in JS it’s always truthy.

jasmussen committed:
Polish the paddings and search field.

This makes the search field a bit bigger and nicer, and changes the height to match better the contents.

jorgefilipecosta committed:
Added test cases and other improvements.

mtias commented on editor/components/inserter/child-items-list.js:
```diff
@@ -0,0 +1,29 @@
+/**
```
This could just be called `ChildBlocks`

jorgefilipecosta replied:
Renamed 👍

gravityrail reviewed (approved):
LGTM! 🎉

gziolo commented on editor/components/inserter/menu.js:
```diff
@@ -176,6 +182,14 @@ export class InserterMenu extends Component {
 
 				<div className="editor-inserter__results">
+					<ChildBlocks
+						rootBlockIcon={ rootBlockIcon }
```
Did you consider to use `withSelect` directly on `ChildBlocks` component instead of passing down `rootBlockIcon` and `rootBlockTitle`?

gziolo commented on editor/components/inserter/child-blocks.js:
```diff
@@ -0,0 +1,29 @@
+import BlockIcon from '../block-icon';
+
+function ChildBlocks( { rootBlockIcon, rootBlockTitle, items, ...props } ) {
+	if ( ! items || ! items.length ) {
```
Can we use `ifCondition` HOC here?

gziolo reviewed (commented):
Code changes look good, I left 2 more questions, but definitely not blockers. Nice work 👍

aduth commented on blocks/api/registration.js:
```diff
@@ -288,3 +288,25 @@ export function hasBlockSupport( nameOrType, feature, defaultSupports ) {
+ * @param {string} blockName Block type name.
+ *
+ * @return {Array} Array of child block names.
+ */
```
Not sure a JSDoc parser would detect this as being "attached" to the `getChildBlockNames` function with the newline between.

jorgefilipecosta replied:
The line breaks were corrected.

aduth commented on editor/components/inserter/item-list.js:
```diff
@@ -74,7 +74,15 @@ class ItemList extends Component {
+									'editor-inserter__item',
+									getBlockMenuDefaultClassName( item.id ),
+									{
+										'editor-inserter__item-with-childs': item.hasChildBlocks,
```
- Grammar: "childs" -> "children"
- This is a modifier, so could be written as one, e.g. `has-children` ([reference](https://github.com/WordPress/gutenberg/blob/master/docs/reference/coding-guidelines.md#naming))

aduth commented on editor/store/selectors.js:
```diff
@@ -1397,6 +1397,7 @@ export const getInserterItems = createSelector(
 				isDisabled,
 				utility: calculateUtility( blockType.category, count, isContextual ),
 				frecency: calculateFrecency( time, count ),
+				hasChildBlocks: hasChildBlocks( blockType.name ),
```
How is the cache bust if we proceed to register a block defining `blockType.name` as a valid `parent` ?

jorgefilipecosta replied:
We clean the cache each time state.blockTypes changes, so when we register a block the cache is invalidated.

aduth replied:
>We clean the cache each time state.blockTypes changes, so when we register a block the cache is invalidated.

But `state.blockTypes` (from `core/blocks` store) is not considered when we call `getInserterItems` (from `core/editor` store).

jorgefilipecosta replied:
I understand now the problem, getInserterItems is cached and depends on core/blocks store when core/blocks change the cache may not be updated. I think this is an already existing problem. Before if we register a new block a new item should exist but the cache may not be updated, it is not easy to spot because the probably the cache updates a lot.

Thank you for clarifying this issue I will try to propose a solution to this problem!

jorgefilipecosta committed:
revisions

jorgefilipecosta commented:
Thank you @jasmussen and @mtias for the design improvements. Thank you @aduth, @gravityrail, and @gziolo for the reviews!

jasmussen commented:
Great work, thank you.

I'm not sure this PR is the cause, but recently I've started seeing duplicate items in the slash command:

<img width="386" alt="screen shot 2018-06-01 at 09 37 11" src="https://user-images.githubusercontent.com/1204802/40827940-7c88f3b0-657f-11e8-8625-0adafb1543a7.png">

Simply search for "/image" and you'll see it. Ring any bells?

aduth commented:
@jasmussen Do you still see this on master? I'm not able to reproduce. Can we create a separate issue if it's still present?

jorgefilipecosta commented:
Sorry, @jasmussen I missed the comment but like @aduth I'm not being able to reproduce the issue in master or in the last release 3.0.1. Are you still able to reproduce the problem?

jasmussen commented:
@aduth @jorgefilipecosta Can confirm that I'm not seeing this anymore. I don't know what fixed it, I could swear I saw it for /gallery as soon as yesterday. But I can't repro.

jasmussen commented:
@aduth @jorgefilipecosta I now see it in `master` again. I have zero idea why it reappeared. And I can't reproduce it for all blocks. I now see it for Heading and Gallery:

![repeat](https://user-images.githubusercontent.com/1204802/41084121-8237b11c-6a33-11e8-8fdc-cd706c53763e.gif)
